<a href="https://colab.research.google.com/github/anujott-codes/SuperStore-Sales-EDA/blob/main/SuperStore_Sales_EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#***Superstore Sales Data: Exploratory Data Analysis***

##1.Data Loading and Initial Exploration

### - Load the dataset into a pandas DataFrame
### - Explore the first few rows to understand the structure of the data
### - Check the data types, summary statistics, and unique values of each column
### - Identify any obvious data quality issues or inconsistencies

In [853]:
# importing all required modules
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import warnings
warnings.filterwarnings('ignore')
from scipy.stats import skew
from scipy.stats import kurtosis

In [854]:
# loading dataset into the DataFrame
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/SuperStore Sales (Case_Study)/Datasets/SuperStore_Dataset.csv")
df

Row ID        Order ID Order Date   Ship Date       Ship Mode  \
0           1  CA-2016-152156   08/11/16    11/11/16    Second Class   
1           2  CA-2016-152156   08/11/16    11/11/16    Second Class   
2           3  CA-2016-138688   12/06/16    16/06/16    Second Class   
3           4  US-2015-108966   11/10/15    18/10/15  Standard Class   
4           5  US-2015-108966   11/10/15    18/10/15  Standard Class   
...       ...             ...        ...         ...             ...   
10009    1048  CA-2016-169103   08/03/16    13/03/16  Standard Class   
10010    9613  CA-2016-110086   18/09/16    22/09/16  Standard Class   
10011    9614  CA-2014-128524   11/11/14  2014-11-13     First Class   
10012    9615  CA-2014-128524   11/11/14  2014-11-13     First Class   
10013    9616  CA-2014-128524   11/11/14  2014-11-13     First Class   

      Customer ID     Customer Name    Segment        Country  \
0        CG-12520       Claire Gute   Consumer  United States   
1        CG-12520       Claire Gute   Consumer  United States   
2        DV-13045   Darrin Van Huff  Corporate  United States   
3        SO-20335    Sean O'Donnell   Consumer  United States   
4        SO-20335    Sean O'Donnell   Consumer  United States   
...           ...               ...        ...            ...   
10009    ZC-21910  Zuschuss Carroll   Consumer  United States   
10010    BD-11320    Bill Donatelli   Consumer  United States   
10011    MZ-17515         Mary Zewe  Corporate  United States   
10012    MZ-17515         Mary Zewe  Corporate  United States   
10013    MZ-17515         Mary Zewe  Corporate  United States   

                  City  ... Postal Code  Region       Product ID  \
0            Henderson  ...       42420   South  FUR-BO-10001798   
1            Henderson  ...       42420   South  FUR-CH-10000454   
2          Los Angeles  ...       90036    West  OFF-LA-10000240   
3      Fort Lauderdale  ...       33311   South  FUR-TA-10000577   
4      Fort Lauderdale  ...       33311   South  OFF-ST-10000760   
...                ...  ...         ...     ...              ...   
10009            Miami  ...       33142   South  FUR-FU-10004006   
10010         Woodland  ...       95695    West  TEC-PH-10001299   
10011          Seattle  ...       98115    West  OFF-AP-10002892   
10012          Seattle  ...       98115    West  TEC-AC-10002718   
10013          Seattle  ...       98115    West  OFF-BI-10003527   

              Category Sub-Category  \
0            Furniture    Bookcases   
1            Furniture       Chairs   
2      Office Supplies       Labels   
3            Furniture       Tables   
4      Office Supplies      Storage   
...                ...          ...   
10009        Furniture  Furnishings   
10010       Technology       Phones   
10011  Office Supplies   Appliances   
10012       Technology  Accessories   
10013  Office Supplies      Binders   

                                            Product Name Sales Price  \
0                      Bush Somerset Collection Bookcase    261.9600   
1      Hon Deluxe Fabric Upholstered Stacking Chairs,...    731.9400   
2      Self-Adhesive Address Labels for Typewriters b...     14.6200   
3          Bretford CR4500 Series Slim Rectangular Table    957.5775   
4                         Eldon Fold 'N Roll Cart System     22.3680   
...                                                  ...         ...   
10009  Deflect-o DuraMat Lighweight, Studded, Beveled...    102.3600   
10010         Polycom CX300 Desktop Phone USB VoIP phone    239.9840   
10011                   Belkin F5C206VTEL 6 Outlet Surge     22.9800   
10012               Belkin Standard 104 key USB Keyboard    102.1300   
10013  Fellowes PB500 Electric Punch Plastic Comb Bin...   2033.5840   

       Quantity  Discount    Profit  
0           2.0      0.00   41.9136  
1           3.0      0.00  219.5820  
2           2.0      0.00    6.8714  
3           5.0      0.45 -383.0310  
4           2.0     

### So our dataset contains 10014 rows/records and has 21 columns.

In [855]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10014 entries, 0 to 10013
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         10014 non-null  int64  
 1   Order ID       10014 non-null  object 
 2   Order Date     10014 non-null  object 
 3   Ship Date      10014 non-null  object 
 4   Ship Mode      9916 non-null   object 
 5   Customer ID    10014 non-null  object 
 6   Customer Name  10014 non-null  object 
 7   Segment        10014 non-null  object 
 8   Country        10014 non-null  object 
 9   City           10014 non-null  object 
 10  State          10014 non-null  object 
 11  Postal Code    10014 non-null  int64  
 12  Region         10014 non-null  object 
 13  Product ID     10014 non-null  object 
 14  Category       10014 non-null  object 
 15  Sub-Category   10014 non-null  object 
 16  Product Name   10014 non-null  object 
 17  Sales Price    10014 non-null  float64
 18  Quanti

### Here we can infer that our dataset has some null values in columns : Ship Mode, Quantity. As well as we can see the datatypes of all the columns so that we can modify them according to our needs.

In [856]:
# statistical overlook of the dataset
df.describe()

Row ID   Postal Code   Sales Price     Quantity      Discount  \
count  10014.000000  10014.000000  10014.000000  9996.000000  10014.000000   
mean    4997.523367  55205.131616    230.038424     3.792517      0.156240   
std     2888.981754  32066.711027    623.057359     2.231021      0.206392   
min        1.000000   1040.000000    -31.500000     1.000000      0.000000   
25%     2494.250000  23223.000000     17.280000     2.000000      0.000000   
50%     4997.500000  57103.000000     54.780000     3.000000      0.200000   
75%     7500.750000  90008.000000    209.970000     5.000000      0.200000   
max     9994.000000  99301.000000  22638.480000    14.000000      0.800000   

             Profit  
count  10014.000000  
mean      28.708317  
std      234.154272  
min    -6599.978000  
25%        1.731375  
50%        8.666500  
75%       29.364000  
max     8399.976000

In [857]:
# various columns in the dataset
df.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State',
       'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category',
       'Product Name', 'Sales Price', 'Quantity', 'Discount', 'Profit'],
      dtype='object')

In [858]:
# Number of unique order ID's i.e unique orders
df['Order ID'].nunique()

5009

### so we have only '5009' unique orders that means our data has alot of duplicate records

In [859]:
# Nulls
print(df.isnull().sum())
print('-'*100)
# Duplicates
print(df.duplicated().sum())


Row ID            0
Order ID          0
Order Date        0
Ship Date         0
Ship Mode        98
Customer ID       0
Customer Name     0
Segment           0
Country           0
City              0
State             0
Postal Code       0
Region            0
Product ID        0
Category          0
Sub-Category      0
Product Name      0
Sales Price       0
Quantity         18
Discount          0
Profit            0
dtype: int64
----------------------------------------------------------------------------------------------------
17


##2. Handling Duplicates

### -Identify and remove duplicate rows in the dataset
### -Document the number of rows and distinct Order IDs affected by this operation


In [860]:
# checking duplicated order ID's
grouped_df = df.groupby('Order ID')['Row ID'].apply(lambda x: x.count()).reset_index()
duplicate_orders = grouped_df[grouped_df['Row ID']>1].reset_index()
print(duplicate_orders)
print('-'*50)
print('Number of duplicate rows in dataset :- ')
print(duplicate_orders['Row ID'].sum())

      index        Order ID  Row ID
0         1  CA-2014-100090       2
1         4  CA-2014-100363       2
2         6  CA-2014-100678       4
3         7  CA-2014-100706       2
4         8  CA-2014-100762       4
...     ...             ...     ...
2471   5001  US-2017-168116       2
2472   5005  US-2017-169320       2
2473   5006  US-2017-169488       2
2474   5007  US-2017-169502       2
2475   5008  US-2017-169551       6

[2476 rows x 3 columns]
--------------------------------------------------
Number of duplicate rows in dataset :- 
7481


###From here we can infer that we have '2476' unique order ID which are repeated multiple times...so keeping only 1 record per order ID and removing the duplicates will remove '7481'-'2476' = '5005' records from the dataset. Hence we can say our dataset had 5005 copy records.

In [861]:
#removing duplicates based on same order id

df.drop_duplicates(subset = 'Order ID',inplace = True)
df

Row ID        Order ID Order Date  Ship Date       Ship Mode  \
0          1  CA-2016-152156   08/11/16   11/11/16    Second Class   
2          3  CA-2016-138688   12/06/16   16/06/16    Second Class   
3          4  US-2015-108966   11/10/15   18/10/15  Standard Class   
5          6  CA-2014-115812   09/06/14  2014-6-14  Standard Class   
12        13  CA-2017-114412   15/04/17   20/04/17  Standard Class   
...      ...             ...        ...        ...             ...   
9986    9987  CA-2016-125794   29/09/16   03/10/16  Standard Class   
9987    9988  CA-2017-163629   17/11/17   21/11/17  Standard Class   
9989    9990  CA-2014-110422   21/01/14  2014-1-23    Second Class   
9990    9991  CA-2017-121258   26/02/17   03/03/17  Standard Class   
9993    9994  CA-2017-119914   04/05/17   09/05/17    Second Class   

     Customer ID     Customer Name    Segment        Country             City  \
0       CG-12520       Claire Gute   Consumer  United States        Henderson   
2       DV-13045   Darrin Van Huff  Corporate  United States      Los Angeles   
3       SO-20335    Sean O'Donnell   Consumer  United States  Fort Lauderdale   
5       BH-11710   Brosina Hoffman   Consumer  United States      Los Angeles   
12      AA-10480      Andrew Allen   Consumer  United States          Concord   
...          ...               ...        ...            ...              ...   
9986    ML-17410      Maris LaWare   Consumer  United States      Los Angeles   
9987    RA-19885      Ruben Ausman  Corporate  United States           Athens   
9989    TB-21400  Tom Boeckenhauer   Consumer  United States            Miami   
9990    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9993    CC-12220      Chris Cortes   Consumer  United States      Westminster   

      ... Postal Code  Region       Product ID         Category Sub-Category  \
0     ...       42420   South  FUR-BO-10001798        Furniture    Bookcases   
2     ...       90036    West  OFF-LA-10000240  Office Supplies       Labels   
3     ...       33311   South  FUR-TA-10000577        Furniture       Tables   
5     ...       90032    West  FUR-FU-10001487        Furniture  Furnishings   
12    ...       28027   South  OFF-PA-10002365  Office Supplies        Paper   
...   ...         ...     ...              ...              ...          ...   
9986  ...       90008    West  TEC-AC-10003399       Technology  Accessories   
9987  ...       30605   South  TEC-AC-10001539       Technology  Accessories   
9989  ...       33180   South  FUR-FU-10001889        Furniture  Furnishings   
9990  ...       92627    West  FUR-FU-10000747        Furniture  Furnishings   
9993  ...       92683    West  OFF-AP-10002684  Office Supplies   Appliances   

                                           Product Name Sales Price  Quantity  \
0                     Bush Somerset Collection Bookcase    261.9600       2.0   
2     Self-Adhesive Address Labels for Typewriters b...     14.6200       2.0   
3         Bretford CR4500 Series Slim Rectangular Table    957.5775       5.0   
5     Eldon Expressions Wood and Plastic Desk Access...     48.8600       7.0   
12                                           Xerox 1967     15.5520       3.0   
...                                                 ...         ...       ...   
9986  Memorex Mini Travel Drive 64 GB USB 2.0 Flash ...     36.2400       1.0   
9987  Logitech G430 Surround Sound Gaming Headset wi...     79.9900       1.0   
9989                             Ultra Door Pull Handle     25.2480       3.0   
9990  Tenex B1-RE Series Chair Mats for Low Pile Car...     91.9600       2.0   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...    243.1600       2.0   

      Discount    Profit  
0         0.00   41.9136  
2         0.00    6.8714  
3         0.45 -383.0310  
5         0.00   14.1694  
12        0.20    5.4432  
...        ...       ...  
9986      0.00   15.2208  
9987      0.00   28.7964  
9989      0.20 

### So clearly now our dataset has '5009' records instead of '10014' records as '5005' records were duplicates and are now removed.

## ~ Distribution Analysis for Numerical columns

In [864]:
cols = ['Sales Price', 'Profit', 'Discount']
for col in cols:
    print(f"{col} Skew: {skew(df[col])}, Kurtosis: {kurtosis(df[col])}")


Sales Price Skew: 9.098371094007186, Kurtosis: 128.20337666682912
Profit Skew: 5.287244104633003, Kurtosis: 232.27159933365868
Discount Skew: 1.6821296658439087, Kurtosis: 2.3785562574537247


### Distributions are right skewed and have heavy tails except for discount which has light tails.

##3. Date Handling

### -Normalize the Order Date and Ship Date columns to ensure consistent date formats. Ensure that the format of the date is consistent across all rows
### -Extract the year from the Order ID and compare it with the year in the Order Date. Correct any inconsistencies
Document the number of rows and distinct Order IDs affected by these operations

In [ ]:
print(df.head())
print('-'*50)
print(df.dtypes)


###So here we can see that Oder Date and Ship Date have object datatype which we be very inconsistent as dates may be stored in different formats -> so we will convert them and normalize them to datetime format so they become consistent.

In [ ]:
df['Order Date'] = pd.to_datetime(df['Order Date'],format='mixed')
df['Ship Date'] = pd.to_datetime(df['Ship Date'],format='mixed')
df

### Now data in both columns i.e Order Date and Ship Date is consistent and has proper datatype.

In [ ]:
# extracting year from Order ID column
df['Order ID Year'] = df['Order ID'].str.split('-').str[1]

#converting to datetime format for comparison
df['Order ID Year'] = pd.to_datetime(df['Order ID Year']).dt.year
df

In [ ]:
# extracting rows with any inconsistencies
df[df['Order ID Year'] != df['Order Date'].dt.year]

### Here we noticed that we have '16' rows where Order ID Year and Order Date year is not same hence we will rectify it and replace Order Date with Order ID year..

In [ ]:
df['Order Date'] = df.apply(lambda row: row['Order Date'].replace(year=int(row['Order ID Year'])), axis=1)

In [ ]:
df[df['Order ID Year'] != df['Order Date'].dt.year]

### So now all the inconsistency in Order Date column is removed...'16' rows were affected from this updation.

In [ ]:
## extra check for Ship Date column
df[df['Order ID Year'] != df['Ship Date'].dt.year]

### So here also we have '72' rows with different years then Order ID but it can be possible as a order can take time to be shipped hence we won't consider it as inconsistent.

In [ ]:
#check for inconsistent data
df[df['Order ID Year'] > df['Ship Date'].dt.year]

### Hence data is consistent and there is no row where shipped Year is lesser than Order ID Year.

##4. Imputation of Missing Values

### - Impute missing values in the Ship Mode column using the calculated Days to Ship column
### - Calculate Days to Ship as the difference between Ship Date and Order Date. If Days to Ship is 0, set Ship Mode to "Same Day"; if it is 7, set Ship Mode to "Standard Class"
### - Impute missing values in the Quantity column using a method of your choice. Print the rationale for selecting the imputation method
### Document the number of rows and distinct Order IDs affected by these operations

### From here we firstly need to make data consistent as in some cases Ship Date Month is earlier than the Order Date Which is not possible

In [ ]:
# firstly to check days to ship

df['Days to Ship'] = (df['Ship Date']-df['Order Date']).dt.days
df

### We can clearly see that some days are negative hence this would lead to inconsistent results. Therefore we need to rectify it

In [ ]:
## checking number of rows with inconsistent data
df[df['Days to Ship']<0]

### clearly a lot of data is inconsistent i.e '853' rows. This is not possible as Order cannot be shipped before it was placed. So we could consider the order was shipped on the same day as it was placed.

In [ ]:
# changing ship date to order date wherever inconsistent
df['Ship Date'] = df.apply(lambda row: row['Order Date']
                           if row['Days to Ship']< 0 else row['Ship Date'],
                           axis = 1)

In [ ]:
#recalculating days

df['Days to Ship'] = (df['Ship Date']-df['Order Date']).dt.days

In [ ]:
## checking for any inconsistency
df[df['Days to Ship']<0]

###Now we can surely say that our data has no inconsistency in Ship Date,Order Date and Days to Ship column...While performing the above operation '853' rows were affected.

In [ ]:
## To classify Ship Mode on basis of average days taken to ship the order

grouped_ship_mode_data = df.groupby('Ship Mode')['Days to Ship'].mean().astype('int').reset_index()
grouped_ship_mode_data

In [ ]:
## to check number of null values in ship mode column

df['Ship Mode'].isna().sum()

###Now imputing the value into Ship Mode column according to mean days taken to ship the order for every mode

In [ ]:
# imputing value into ship mode column on the basis of above mean days to ship data
conditions = [
    df['Ship Mode'].isna() & (df['Days to Ship'] <= 6),
    df['Ship Mode'].isna() & (df['Days to Ship'] > 6) & (df['Days to Ship'] <= 28),
    df['Ship Mode'].isna() & (df['Days to Ship'] > 28) & (df['Days to Ship'] <= 34),
    df['Ship Mode'].isna() & (df['Days to Ship'] > 34)
]

choices = ['Same Day', 'First Class', 'Second Class', 'Standard Class']

df['Ship Mode'] = np.select(conditions, choices, default=df['Ship Mode'])

In [ ]:
## checking if any null value remains
df['Ship Mode'].isna().sum()
df['Ship Mode'].isna().any()

### Hence now we can say that there is no inconsistency in our ship mode column and in total '29' records were altered during this operation...

In [ ]:
## to check number of null values in quantity column

df['Quantity'].isna().sum()

In [ ]:
df[df['Quantity'].isna()]

In [ ]:
#checking the distribution of Quantity column
sns.histplot(df['Quantity'], kde=True)
print("Skewness:", skew(df['Quantity']))


###So as Quantity is right skewed we would prefer median for filling NaN values

In [ ]:
# imputing mean values based on sub-category grouped data
grp_sub_category_data = df.groupby('Sub-Category')['Quantity'].median().astype('int').reset_index()
grp_sub_category_data

In [ ]:
grp_sub_category_data['Quantity'].nunique()

### So basically I decided to fill NaN values for quantity on basis of median quantity value sub-category wise and on average median is very close to 3 so lets fill NaN values with 3.

In [ ]:
df['Quantity'] = df['Quantity'].fillna(3)

In [ ]:
# checking if any missing value still exists

df['Quantity'].isna().sum()

In [ ]:
df['Quantity'].isna().any()

### Hence now we can say that all the missing values i.e '8' have been treated properly

##5. Data Masking and String Handling

### - Drop the Customer Name column to protect Personal Identifiable Information (PII)
### - Create a new column called Customer Name Masked, containing only the initials of the customer name

### ~ Note: It's important to protect PII in datasets to maintain customer privacy and comply with data protection regulations. Masking or dropping sensitive data like customer names is a crucial step in this process
### - Convert the Postal Code column from numeric to text format, ensuring all codes are 5 characters long. Add a leading '0' where necessary

In [ ]:
# Masking names of customer with there initials
df['Customer Name Masked'] = df['Customer Name'].apply(lambda name: (name.split(" ")[0][0] + name.split(" ")[1][0]) if len(name.split(" ")) >= 2 else name.split(" ")[0])

In [ ]:
#dropping Customer Name column to protect PII
df.drop(columns=['Customer Name'],inplace=True)

In [ ]:
## rechecking dataframe
print('Customer Name' in df.columns)
print('='*200)
df

### So by now we have successfully removed the customer name column to protect PII and created a new Customer Name Mask column

In [ ]:
df['Postal Code'].dtypes

In [ ]:
## converting postal code to text from numeric

df['Postal Code'] = df['Postal Code'].astype('str')

In [ ]:
df['Postal Code'].dtypes

In [ ]:
#Updating Postal Code wherever required
df['Postal Code'] = df['Postal Code'].apply(lambda x: "0"*(5-len(x))+x if len(x) < 5 else x)

### Now all the postal codes are converted in text from numeric and are of length 5...leading zeroes are added wherever required

##6. Data Type Conversion

### - Convert the Quantity and Sales Price columns from strings to their appropriate numeric types (int and float, respectively)

In [ ]:
df['Quantity'] = df['Quantity'].astype('int')
df['Sales Price'] = df['Sales Price'].astype('float')

In [ ]:
df['Quantity'].dtype

In [ ]:
df['Sales Price'].dtype

###The columns Quantity and Sales Price are converted to their respective data types i.e [int , float]

##7. Handling Inconsistent Categorical Data

###Clean the State column by replacing abbreviations with full state names (e.g., "CA" should be changed to "California"). You may need to research state abbreviations online to ensure all entries are corrected consistently

In [ ]:
df['State']

In [ ]:
states = [
    "Alabama", "Alaska", "Arizona", "Arkansas", "American Samoa",
    "California", "Colorado", "Connecticut", "Delaware", "District of Columbia",
    "Florida", "Georgia", "Guam", "Hawaii", "Idaho",
    "Illinois", "Indiana", "Iowa", "Kansas", "Kentucky",
    "Louisiana", "Maine", "Maryland", "Massachusetts", "Michigan",
    "Minnesota", "Mississippi", "Missouri", "Montana", "Nebraska",
    "Nevada", "New Hampshire", "New Jersey", "New Mexico", "New York",
    "North Carolina", "North Dakota", "Northern Mariana Islands", "Ohio", "Oklahoma",
    "Oregon", "Pennsylvania", "Puerto Rico", "Rhode Island", "South Carolina",
    "South Dakota", "Tennessee", "Texas", "Trust Territories", "Utah",
    "Vermont", "Virginia", "Virgin Islands", "Washington", "West Virginia",
    "Wisconsin", "Wyoming"
]
state_codes = [
    "AL", "AK", "AZ", "AR", "AS",
    "CA", "CO", "CT", "DE", "DC",
    "FL", "GA", "GU", "HI", "ID",
    "IL", "IN", "IA", "KS", "KY",
    "LA", "ME", "MD", "MA", "MI",
    "MN", "MS", "MO", "MT", "NE",
    "NV", "NH", "NJ", "NM", "NY",
    "NC", "ND", "MP", "OH", "OK",
    "OR", "PA", "PR", "RI", "SC",
    "SD", "TN", "TX", "TT", "UT",
    "VT", "VA", "VI", "WA", "WV",
    "WI", "WY"
]
df[df['State'].isin(state_codes)]

In [ ]:
code_vs_state = {code:state for code,state in zip(state_codes,states)}
code_vs_state

In [ ]:
df['State'].unique()
df['State'] = df['State'].str.strip(' /\\')
df['State'] = df['State'].apply(lambda x: code_vs_state[x] if x in state_codes else x)

In [ ]:
df[df['State'].isin(state_codes)]

### Hence all the state codes are now converted into respective Country names...A total of '58' rows were affected during this operation

##8. Feature Engineering

##Create new columns
###-Original Price: The price before any discount is applied
###-Total Sales: The total revenue is generated by multiplying the Sales Price by Quantity
###-Total Profit: The total profit is earned by multiplying the Profit by Quantity
###-Discount Price: The amount of discount applied, calculated based on the Original Price and Discount
###-Total Discount: The total discount value for the quantity sold


In [ ]:
df

In [ ]:
## creating new column for Original price that is price before any discount was given
## formula used => x = y/(1-d)
df['Original Price'] = (df['Sales Price']/(1-df['Discount']))
df

In [ ]:
# creating new column for total revenue i.e sales price * quantity
df["Total Revenue"] = (df['Sales Price'] * df['Quantity'])
df

In [ ]:
# creating new column total profit i.e profit * quantity
df['Total Profit'] = (df['Profit'] * df['Quantity'])
df

In [ ]:
#creating new column discount price i.e amount of discount given
#formula => dicount price = original price * discount
df['Discount Price'] = (df['Original Price'] * df['Discount'])
df

In [ ]:
# creating new column total discount i.e discount price * quantity
df["Total Discount"] = df["Discount Price"] * df["Quantity"]
df



## ~  Create a new column, Shipping Urgency, based on Days to Ship

 ### -If Days to Ship is 0, set to "Immediate"

 ### -If Days to Ship is between 1 and 3, set to "Urgent"

 ### -If Days to Ship is more than 3, set to "Standard"




In [ ]:
#creating new column shipping urgency based on above conditons

df['Shipping Urgency'] = 0

conditions = [
    df['Days to Ship'] == 0,
    (df['Days to Ship'] >0) & (df['Days to Ship'] <= 3),
    df['Days to Ship'] > 3
               ]
choices = ['Immediate','Urgent','Standard']
df['Shipping Urgency'] = np.select(conditions,choices,default = df['Shipping Urgency'])

In [ ]:
df[['Days to Ship','Shipping Urgency']]

In [ ]:
from datetime import datetime
df['Order Date'] = pd.to_datetime(df['Order Date'])
df['Days Last Order'] = (datetime.now()-df['Order Date']).dt.days
df

In [ ]:
temp = df.groupby('Customer ID')['Days Last Order'].min().reset_index()
temp

In [ ]:
days_map =  dict(zip(temp['Customer ID'],temp['Days Last Order']))
df['Days Last Order'] = df['Customer ID'].map(days_map)

In [ ]:
df

### Now we have successfully added days since last order for every unique customer

## ~Create a new dataset that stores the total sales, quantity, and discount per customer, and then merge these back to the original dataset


In [ ]:
df_customers = df.groupby('Customer ID')[['Total Revenue','Quantity','Discount']].sum().reset_index()
df_customers

In [ ]:
df = pd.merge(df,df_customers,on='Customer ID',how='inner')
df

In [ ]:
df.isna().any()

In [ ]:
df.rename(columns={'Total Revenue_y':'Overall Revenue Generated','Quantity_y':'Overall Quantity Sold','Discount_y':'Overall Discount Given','Discount_x':'Discount','Quantity_x':'Quantity','Total Revenue_x':'Total Revenue'},inplace=True)
df

##9. Outlier Detection and Handling

### - Identify and handle outliers in the Sales Price column
### - Create a function called remove_outliers that takes as an argument the dataframe and the column that needs to be searched for outliers. Using the 3 * IQR rule, the function should detect and remove the outliers to return the cleaned dataframe
### - Now use the function to detect outliers and remove outliers from the Sales Price and Profit columns


In [ ]:
#distribution of sales price before cleaning
sns.boxplot(x=df['Sales Price'])
plt.title('Sales Price Distribution (Before Cleaning)')
plt.xlabel('Sales Price')
plt.grid(True)
plt.show()

###Clearly we can see that sales price has too many outliers

In [ ]:
## function definition

def remove_outliers(df,col):
  q3 = df[col].quantile(0.75)       # 3rd quantile
  q1 = df[col].quantile(0.25)       # 1st quantile
  iqr = (q3-q1)                     # IQR

  lower_bound = q1 - (3 * iqr)
  upper_bound = q3 + (3 * iqr)

  df = df[(df[col]>lower_bound) & (df[col]<upper_bound)]
  return df


In [ ]:
# function call for Sales Price
df = remove_outliers(df,'Sales Price')

In [ ]:
# function call for Profit
df = remove_outliers(df,'Profit')

In [ ]:
#Distribution of sales price after cleaning
sns.boxplot(x=df['Sales Price'])
plt.title('Sales Price Distribution (Before Cleaning)')
plt.xlabel('Sales Price')
plt.grid(True)
plt.show()

In [ ]:
## final df after removing outliers
df

### Earlier we had 5009 rows now only 4212, hence 797 rows were affected from this operations.

##10. Customer Segmentation and Analysis

### - Calculate Customer Sales Quintile and Customer Profit Quintile based on total sales and total profit per Customer ID
### - What is a Quintile? Quintiles are a statistical way of dividing data into five equal parts, each representing 20% of the data. For example, customers in the top quintile (Q5) represent the top 20% of sales or profit

In [ ]:
customer_df = df.groupby('Customer ID')[['Total Revenue','Total Profit']].sum().reset_index()
customer_df.rename(columns={'Total Revenue':'Total Sales'},inplace=True)
customer_df

In [ ]:
## making Sales Quintile
customer_df['Sales Quintile'] = pd.qcut(customer_df['Total Sales'],5,labels=[1,2,3,4,5])

## making Profit Quintile
customer_df['Profit Quintile'] = pd.qcut(customer_df['Total Profit'],5,labels=[1,2,3,4,5])

customer_df

##Create a cross-grid (cross-tabulation) based on these two quintiles to analyze the relationship between customer sales and profitability

In [ ]:
cross_relation_matrix = pd.crosstab(customer_df['Sales Quintile'],customer_df['Profit Quintile'])
cross_relation_matrix

##11.Final Analysis and Dashboard Creation

## ~ Sales and Profit Analysis

###  - Top 10 Most Profitable Products: Use a bar chart to display the products with the highest total profit


In [ ]:
top_products = df.sort_values(by='Total Profit',ascending = False)[['Product Name','Total Profit']][:10]
top_products

In [ ]:
##plotting bar chart for top products and their profit
plt.figure(figsize=(8,8))
sns.barplot(x='Product Name',y='Total Profit',data = top_products,palette='viridis',hue='Product Name')
plt.title('Top 10 Most Profitable Products')
plt.xlabel("Product Name")
plt.ylabel("Profit Earned")
plt.xticks(rotation = 90)
plt.grid(axis='y',linestyle='--',color='k')
plt.show()

###- Top 10 Most Loss-Making Products: Use a bar chart to display the products with the highest total losses (negative profit)

In [ ]:
## checking for most loss-making products
bottom_products = df.sort_values(by='Total Profit',ascending=True)[['Product Name','Total Profit']].head(10)
bottom_products["Total Profit"] = bottom_products["Total Profit"].abs()
bottom_products.rename(columns={'Total Profit':'Total Loss'},inplace=True)
bottom_products

In [ ]:
##plotting bar chart for most lost mmaking products and their loss
plt.figure(figsize=(8,8))
sns.barplot(x='Product Name',y='Total Loss',data = bottom_products,palette='plasma',hue='Product Name')
plt.title('Top 10 Most Lost Making Products')
plt.xlabel("Product Name")
plt.ylabel("Loss Made")
plt.xticks(rotation = 90)
plt.grid(axis='y',linestyle='--',color='k')
plt.show()

### - Sales vs. Profit Correlation: Use a scatter plot to visualize the correlation between Total Sales and Total Profit. Add a regression line to show the trend

In [ ]:
### plotting Total Sales vs Total Profit Customer wise
customer_df

In [ ]:
plt.figure(figsize=(8,8))
sns.scatterplot(x='Total Sales',y='Total Profit',data = customer_df)
sns.regplot(x='Total Sales',y='Total Profit',data = customer_df,color = 'red',scatter = False,line_kws={'linestyle': '--'})
plt.title('Customer-Wise Total Sales vs. Total Profit with Regression Line')
plt.xlabel('Total Sales')
plt.ylabel('Total Profit')
plt.grid(True, linestyle='--', alpha=0.6)
plt.show()

### ~ Joint Distribution of Sales and Profit: Create a joint plot to analyze the relationship between Total Sales and Total Profit across different products

In [ ]:
## firstly we need to create a product_df to analyse total sales and total profit
product_df = df.groupby('Product Name')[['Total Revenue','Total Profit']].sum().reset_index()
product_df

In [ ]:
## plotting the joint plot on basis of above data
plt.figure(figsize=(8,6))
sns.jointplot(x='Total Revenue',y='Total Profit',data = product_df)
plt.title('Joint Distribution of Sales and Profit')
plt.xlabel('Total Sales Product wise')
plt.ylabel('Total Profit')
plt.tight_layout()
plt.show()

##Customer Segmentation and Analysis

### - Customer Sales Quintile vs. Customer Profit Quintile: Create a heatmap or cross-tabulation to explore the relationship between customers' sales and profit quintiles. This will help identify which segments of customers are most valuable

In [ ]:
## heatmap for sales quintile and profit quintile
plt.figure(figsize = (8,6))
sns.heatmap(cross_relation_matrix,annot=True,linewidth=0.5)
plt.title("Heatmap for Sales Quintile and Profit Quintile")
plt.xlabel("Profit Quintile")
plt.ylabel("Sales Quintile")
plt.show()

## ~ Understand how different product categories perform across customer segments: Create a pivot table to analyze the total Sales and total Profit by Category and Segment. Sort the pivot table to highlight the most profitable and least profitable combinations of Category and Segment


In [ ]:
pivot_df = df.pivot_table(index='Segment',columns='Category',values=['Total Revenue','Total Profit'],aggfunc='sum').reset_index()
pivot_df

In [ ]:
profit_df = pivot_df['Total Profit'].copy()
profit_df['Segment'] = pivot_df['Segment']
profit_df

In [ ]:
profit_updated = pd.melt(
    profit_df,
    id_vars='Segment',
    var_name='Category',
    value_name='Total_Profit'
)

profit_updated = profit_updated.sort_values(by=['Segment','Total_Profit'],ascending=[True,False])
profit_updated

In [ ]:
revenue_df = pivot_df['Total Revenue'].copy()
revenue_df['Segment'] = pivot_df['Segment']
revenue_df

In [ ]:
revenue_updated = pd.melt(
    revenue_df,
    id_vars='Segment',
    var_name = 'Category',
    value_name= 'Total_Revenue'
)
revenue_updated = revenue_updated.sort_values(by=['Segment','Total_Revenue'],ascending=[True,False])
revenue_updated

##So clearly we can say the most profitable combinations are [(Consumer - Office Supplies),(Corporate - Office Supplies),(Home Office - Office Supplies)] and the least profit making combinations are [(consumer - Technology), (Corporate - Technology), (Home Office - Technology)]



#*  Shipping and Delivery Analysis



## ~ Distribution of Shipping Urgency: Visualize the distribution of orders by Shipping Urgency using a pie chart or bar chart

In [ ]:
data = df.groupby('Shipping Urgency')['Row ID'].count().reset_index()
data

In [ ]:
# extracting values and labels from data
values = data['Row ID']
labels = data['Shipping Urgency']
explode = [0.01,0.01,0.01]

# plotting a pie chart for order distribution
plt.figure(figsize=(6,6))
plt.pie(values,labels=labels,autopct='%1.1f%%',explode=explode)
plt.title('Pie Chart for Distribution of Shipping Urgency')
plt.axis('equal')
plt.show()

In [ ]:
#plotting a bar chart for order distribution
sns.barplot(data=data,x='Shipping Urgency',y='Row ID',palette='viridis',hue='Shipping Urgency')
plt.title('Bar Chart for Distribution of Shipping Urgency')
plt.xlabel('Shipping Urgency')
plt.ylabel('Distribution')
plt.show()

## ~ Days to Ship vs. Profit: Use a violin plot to explore the distribution of Profit across different Days to Ship categories. This will help analyze whether faster shipping correlates with higher or lower profitability


In [ ]:
#need to divide Days to Ship into categories i.e ship category
bins = [0, 5, 30, float('inf')]
labels = ['1-5 Days', '6-30 Days', '31+ Days']
df['Ship Category'] = pd.cut(df['Days to Ship'], bins=bins, labels=labels)

#Plotting violinplot
plt.figure(figsize=(8, 6))
sns.violinplot(data=df, x='Ship Category', y='Total Profit')
plt.title('Profit Distribution by Shipping Category')
plt.xlabel('Days taken for Shipping')
plt.ylabel('Profit Earned')
plt.grid(axis='y', linestyle='--', alpha=0.5)
plt.tight_layout()
plt.show()

### The violin plot suggests that profit distribution does not depend much on shipping days as it is almost same for the all three categories

## ~ Shipping Mode and Profitability: Create a grouped bar chart to compare the profitability of different shipping modes (e.g., Standard Class, First Class)

In [ ]:
# grouping data on basis of shipping mode and analysing total profit per mode
## we will use avg profit as number of orders per class may vary hence profit will vary significantly
grouped_df = df.groupby('Ship Mode').agg({'Total Profit':'sum','Row ID':'count'}).reset_index()
grouped_df['Avg Profit'] = round(grouped_df['Total Profit']/grouped_df['Row ID'],2)
grouped_df

## plotting the bar chart for average profit per shipping mode
sns.barplot(data=grouped_df,x='Ship Mode',y='Avg Profit')
plt.title('Profitability across each Shipping Mode')
plt.xlabel('Shipping Mode')
plt.ylabel('Profit Earned')
plt.grid(axis='y',linestyle='--',alpha=0.7,color='r')
plt.tight_layout()
plt.show()

## plotting the bar chart for total profit per shipping mode
sns.barplot(data=grouped_df,x='Ship Mode',y='Total Profit')
plt.title('Profitability across each Shipping Mode')
plt.xlabel('Shipping Mode')
plt.ylabel('Profit Earned')
plt.grid(axis='y',linestyle='--',alpha=0.7,color='r')
plt.tight_layout()
plt.show()

## Here as we observed there is not much significant difference between average profit earned within each shipping mode, still "Second Class" is most profitable amongst all, but if we plot total profit for each shipping mode there will be a significant difference as number of orders per shipping mode vary significantly

### We can visualise profit for each shipping mode sub categorised by segment

In [ ]:
#preparing data for plotting avg profit per segment in each shipping moderrddfdc vfd
grouped_df = df.groupby(['Ship Mode', 'Segment']).agg({'Total Profit': 'mean'}).reset_index()
grouped_df['Total Profit'] = grouped_df['Total Profit'].round(2)

# plotting grouped bar chart
plt.figure(figsize=(10, 6))
sns.barplot(data=grouped_df, x='Ship Mode', y='Total Profit', hue='Segment', palette='plasma')

plt.title('Average Profit by Shipping Mode and Segment')
plt.xlabel('Shipping Mode')
plt.ylabel('Average Profit')
plt.grid(axis='y', linestyle='--', alpha=0.7)
plt.tight_layout()
plt.show()

## ~ Using a pivot table, determine which shipping modes are most preferred across different regions and analyze the impact on total sales and profit. Create a pivot table that shows the count of Order IDs, total Sales, and total Profit for each Region and Ship Mode. Identify and print your insights

In [ ]:
pivot_region = df.pivot_table(index = 'Region',columns='Ship Mode',values=['Order ID','Total Revenue','Total Profit'],aggfunc={'Order ID':'count','Total Revenue':'sum','Total Profit':'sum'}).reset_index()
pivot_region

In [ ]:
most_prefered = pivot_region['Order ID'].copy()
most_prefered['Region'] = pivot_region['Region']
most_prefered = pd.melt(most_prefered,
                        id_vars='Region',
                        var_name='Ship Mode',
                        value_name='Num_of_Orders'
                        )
most_prefered.groupby(['Region'])[['Ship Mode','Num_of_Orders']].max().reset_index()


### So clearly we can see that Standard Class is most preferred amongst all the Regions

In [ ]:
# analysing profit region wise
most_profitable = pivot_region['Total Profit'].copy()
most_profitable['Region'] = pivot_region['Region']
most_profitable = pd.melt(most_profitable,
                        id_vars='Region',
                        var_name='Ship Mode',
                        value_name='Total Profit'
                        )
most_profitable.groupby(['Region'])[['Ship Mode','Total Profit']].max().reset_index()

###So according to our previous analysis it was expected that standard class would be most profitable because it had maximum number of orders per region

In [ ]:
#To analyse average profit per region
most_prefered_1 = pivot_region['Order ID'].copy()
most_prefered_1['Region'] = pivot_region['Region']
most_prefered_1 = pd.melt(most_prefered_1,
                        id_vars='Region',
                        var_name='Ship Mode',
                        value_name='Num_of_Orders'
                        )
most_profitable_1 = pivot_region['Total Profit'].copy()
most_profitable_1['Region'] = pivot_region['Region']
most_profitable_1 = pd.melt(most_profitable_1,
                        id_vars='Region',
                        var_name='Ship Mode',
                        value_name='Total Profit'
                        )
final_df = pd.merge(most_profitable_1,most_prefered_1,on=['Region','Ship Mode'],how='inner')
final_df['Avg Profit'] = final_df['Total Profit']/final_df['Num_of_Orders']
final_df.groupby('Region')[['Ship Mode','Avg Profit']].max().reset_index()

###So now we can finally compute and say that Standard Class is the most profitable class across all regions...

In [ ]:
# analysing total sales i.e total revenue earned
most_sales = pivot_region['Total Revenue'].copy()
most_sales['Region'] = pivot_region['Region']
most_sales = pd.melt(most_sales,
                        id_vars='Region',
                        var_name='Ship Mode',
                        value_name='Total Sales'
                        )
most_sales.groupby(['Region'])[['Ship Mode','Total Sales']].max().reset_index()

### As expected from our above analysis Standard Class has generated the maximum sales across all regions

## ~ Regional Sales and Profitability

### - Sales and Profit by Region: Use a map or bar chart to visualize total sales and profit by region or state. This will highlight which regions are the most profitable

In [ ]:
#visualising total sales and total profit region wise
#firstly total sales
plt.figure(figsize=(8,6))
sns.barplot(data=most_sales,x='Region',y='Total Sales',hue='Ship Mode',palette='viridis')
plt.title('Total Sales per Region')
plt.tight_layout()
plt.show()

###It's very evident that Standard Class has the maximum sales per region as analysed before

In [ ]:
#now total profit per region
plt.figure(figsize=(8,6))
sns.barplot(data=most_profitable,x='Region',y='Total Profit',hue='Ship Mode',palette='inferno')
plt.title('Total Profit per Region')
plt.tight_layout()
plt.show()

###So clearly standard class is most profitable across all regions

## ~ State-wise Profitability: Create a pivot table to summarize the profitability of each state. Highlight the top and bottom states based on profitability

In [ ]:
# pivot_state = df.pivot_table(index='State',values='Total Profit',aggfunc='mean')
#  similar but easier to understand
pivot_state = df.groupby('State')['Total Profit'].mean().reset_index()
pivot_state = pivot_state.sort_values(by='Total Profit',ascending=False)
pivot_state.rename(columns={'Total Profit':"Average Profit"},inplace=True)
pivot_state

In [ ]:
# top 5 profitable states
top_5_states = pivot_state.head(5)

# top 5 least profitable states
bottom_5_states = pivot_state.tail(5)

print("Top 5 Profitabe states are:-\n")
print(top_5_states)
print('-'*200)
print('Top 5 Least Profitabe states are:-\n')
print(bottom_5_states)

###Here we can easily analyse our top 5 most profitable states and least profitable states where (-ve) sign refers to loss in a particular state

## ~ Correlation between State and Profit: Use a correlation plot to identify any patterns or relationships between the states and the profitability of orders

In [ ]:
#To find a proper correlation matrix between states and profit we need to conver states in numerical column

#importing LabelEncoder
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

#Encoding State Column
pivot_state['State_encoded'] = le.fit_transform(pivot_state['State'])

#Preparing correlation matrix
correlation_matrix = pivot_state[['State_encoded','Average Profit']].corr()

#plotting the correlation_matrix
plt.figure(figsize=(6,4))
sns.heatmap(correlation_matrix, annot=True, cmap='plasma', fmt=".2f")
plt.title('Correlation between State and Average Profit')
plt.show()



###we can infer that the state does not show a strong linear trend in influencing profit. But as an individual some states are profitable and some states are loss bearing and we can visualise that through a bar chart below

In [ ]:
#plotting barplot
plt.figure(figsize=(10, 12))
sns.barplot(data=pivot_state,x='State',y='Average Profit',palette ='viridis',hue='State')
plt.axhline(pivot_state['Average Profit'].mean(),color='red',linestyle='--',label='mean')
plt.title('Average Profit by State')
plt.xlabel('States')
plt.ylabel('Average Profit')
plt.xticks(rotation=90)
plt.legend()
plt.tight_layout()
plt.show()

### As suggested from above heatmap some states are highly profitable like [Vermont,West Virginia,Mississippi,Virginia, New Hampshire, Indiana, Alabama ] and some of the states are not so profitable or loss bearing like [Illinois, Florida, Tennessee, Texas, Ohio]

## ~ Discount and Pricing Analysis

### -> Impact of Discounts on Profitability: Use a scatter plot with a trend line to analyze how different levels of discount affect profitability

In [ ]:
#plotting a regplot with scatter and regression line
plt.figure(figsize=(8,6))
sns.regplot(data=df, x='Discount', y='Total Profit', scatter_kws={'alpha':0.5}, line_kws={'color':'red','linestyle':'--'})
plt.title('Impact of Discount on Profit')
plt.xlabel('Discount (%)')
plt.ylabel('Total Profit')
plt.grid(True)
plt.show()


###Clearly the graph depicts a negative correlation between discount and profit.Increasing the discount to very high margins such as 0.7 or 0.8 leads to loss bearing orders and hence we can infer that giving more discount may generate more sales but not always generate more profit.

## ~ Original Price vs. Discounted Price: Create a line plot to compare the original price and the discounted price across various product categories or sub-categories


In [ ]:
# creating a column discounted price i.e original price - discount price
df['Discounted Price'] = df['Original Price'] - df['Discount Price']
df[['Original Price','Discounted Price']]

In [ ]:
#prepare data for original price and discounted price category wise
data = df.groupby('Category')[['Original Price','Discounted Price']].mean().reset_index()

#plot graph
plt.figure(figsize=(6,4))
plt.plot(data['Category'],data['Original Price'],color='green',marker='D',label='Original Price')
plt.plot(data['Category'],data['Discounted Price'],color='red',marker='D',label='Discounted Price')
plt.title('Comparison between Original Price and Discounted Price Category wise')
plt.xlabel('Categories')
plt.ylabel('Prices')
plt.xticks(rotation=45)
plt.show()

###We can infer that Average Prices for Office Supplies is comparitively very less than that of Furniture and Technology

In [ ]:
#prepare data for original price and discounted price sub-category wise
data = df.groupby('Sub-Category')[['Original Price','Discounted Price']].mean().reset_index()

#plot graph
plt.figure(figsize=(6,4))
plt.plot(data['Sub-Category'],data['Original Price'],color='green',label='Original Price')
plt.plot(data['Sub-Category'],data['Discounted Price'],color='red',label='Discounted Price',linestyle='--')
plt.title('Comparison between Original Price and Discounted Price Sub-Category wise')
plt.xlabel('Sub-Categories')
plt.ylabel('Prices')
plt.xticks(rotation=90)
plt.legend()
plt.grid(True,linestyle='--',color='gray',alpha=0.5,linewidth=0.7)
plt.show()

###Here we can infer that heavy discounts were given in sub-categories like 'Bookcases','Chairs','Copiers','Machines', and very less discount is given in sub-categories like 'Art','Envelopes','Fasteners','Labels','Paper',etc. this difference may be due to high ad low margins.

In [ ]:
# analysing trends within sub-categories of a common category
data = df.groupby(['Category','Sub-Category'])[['Original Price','Discounted Price']].mean().reset_index()
furniture_df = data[data['Category']=='Furniture']
office_df = data[data['Category']=='Office Supplies']
tech_df = data[data['Category']=='Technology']

In [ ]:
# Set index to Sub-Category
furniture_df.set_index('Sub-Category', inplace=True)

# Plot
plt.figure(figsize=(10,6))
plt.plot(furniture_df.index, furniture_df['Original Price'], marker='o', label='Original Price', linewidth=2)
plt.plot(furniture_df.index, furniture_df['Discounted Price'], marker='o', label='Discounted Price', linewidth=2, linestyle='--')
plt.title('Original vs Discounted Price by Furniture-Sub-Category')
plt.xlabel('Sub-Category')
plt.ylabel('Price')
plt.xticks(rotation=45)
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()


In [ ]:
# Set index to Sub-Category
office_df.set_index('Sub-Category', inplace=True)

# Plot
plt.figure(figsize=(10,6))
plt.plot(office_df.index, office_df['Original Price'], marker='o', label='Original Price', linewidth=2)
plt.plot(office_df.index, office_df['Discounted Price'], marker='o', label='Discounted Price', linewidth=2, linestyle='--')
plt.title('Original vs Discounted Price by Office Supplies-Sub-Category')
plt.xlabel('Sub-Category')
plt.ylabel('Price')
plt.xticks(rotation=45)
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()

In [ ]:
# Set index to Sub-Category
tech_df.set_index('Sub-Category', inplace=True)

# Plot
plt.figure(figsize=(10,6))
plt.plot(tech_df.index, tech_df['Original Price'], marker='o', label='Original Price', linewidth=2)
plt.plot(tech_df.index, tech_df['Discounted Price'], marker='o', label='Discounted Price', linewidth=2, linestyle='--')
plt.title('Original vs Discounted Price by Technology-Sub-Category')
plt.xlabel('Sub-Category')
plt.ylabel('Price')
plt.xticks(rotation=45)
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()

###Now we have completely analysed original price and discounted price for our each category and sub-categories within each category

## ~ Sales and Profit Trends Over Time: Use a time series plot to analyze how sales and profit have trended over the years or months. This will help in identifying any seasonal patterns

In [ ]:
#convert data into year wise periods
yearly_df = df.groupby([df['Order Date'].dt.to_period('Y')])[['Total Revenue','Total Profit']].sum().reset_index()
yearly_df['Order Date'] = yearly_df['Order Date'].astype('str')

#plot the graph
plt.figure(figsize=(8,6))
plt.plot(yearly_df['Order Date'],yearly_df['Total Revenue'],label='Total Sales',color='blue',marker='o')
plt.plot(yearly_df['Order Date'],yearly_df['Total Profit'],label='Total Profit',color='red',marker='*')
plt.title('Series Time plot for Annual Total Sales and Total Profit')
plt.xlabel('Years')
plt.ylabel('Amount')
plt.legend()
plt.grid(True,linestyle='--',alpha=0.5,color='gray',linewidth=0.5)
plt.show()

###This plot shows that annual sales are very high but corresponding annual profit margin is not that high.
###Sales and Profit have increased gradually over 2015-2017


In [ ]:
#convert data into month wise periods
monthly_df = df.groupby([df['Order Date'].dt.to_period('M')])[['Total Revenue','Total Profit']].sum().reset_index()
monthly_df['Order Date'] = monthly_df['Order Date'].astype('str')

#plot the graph
plt.figure(figsize=(14,8))
plt.plot(monthly_df['Order Date'],monthly_df['Total Revenue'],label='Total Sales',color='blue',marker='o')
plt.plot(monthly_df['Order Date'],monthly_df['Total Profit'],label='Total Profit',color='red',marker='*')
plt.title('Series Time plot for Monthly Total Sales and Total Profit')
plt.xlabel('Month')
plt.ylabel('Amount')
plt.xticks(rotation=90)
plt.legend()
plt.grid(True,linestyle='--',alpha=0.5,color='gray',linewidth=0.5)
plt.show()

###Here we can notice a seasonal spike in sales and profit near the end of the year i.e in the months from october to december...then a sudden decrease near the january month...

## ~ Order Frequency by Month: Use a bar chart or line plot to show the number of orders placed each month. Highlight any months with unusually high or low order frequencies

In [ ]:
# Prepare monthly orders count data
orders_per_month = df.groupby(df['Order Date'].dt.to_period('M'))['Order ID'].count().reset_index()
orders_per_month.rename(columns={'Order ID': 'Num of Orders'}, inplace=True)
orders_per_month['Order Date'] = orders_per_month['Order Date'].astype(str)  # Convert Period to string for plotting

# Plot a bar graph with a line plot
fig, ax = plt.subplots(figsize=(14, 8))
sns.barplot(data=orders_per_month, x='Order Date', y='Num of Orders', palette='viridis', ax=ax)
sns.lineplot(data=orders_per_month, x='Order Date', y='Num of Orders', color='red',linestyle='--', marker='o', ax=ax)

ax.set_title('Number of Orders Placed per Month')
ax.set_xlabel('Order Date')
ax.set_ylabel('Num of Orders')
plt.xticks(rotation=90)
plt.tight_layout()
plt.show()


###We can observe that number of orders has gradually increased over time with some seasonal rise in specific months like september,november,december

## ~ Yearly Growth in Sales and Profit: Use a year-over-year growth chart to compare the sales and profit growth over different years



In [ ]:
yearly_df['Profit Growth'] = yearly_df['Total Profit'].pct_change() * 100
yearly_df['Sales Growth'] = yearly_df['Total Revenue'].pct_change() * 100
yearly_df.fillna(0,inplace=True)
yearly_df

In [ ]:
#plotting year over year growth in Sales and Profit
plt.figure(figsize=(10,6))
plt.plot(yearly_df['Order Date'], yearly_df['Sales Growth'], marker='o', label='Sales Growth')
plt.plot(yearly_df['Order Date'], yearly_df['Profit Growth'], marker='o', label='Profit Growth')
plt.axhline(0, color='red',alpha=0.5)
plt.title('Year-over-Year Growth in Sales and Profit')
plt.xlabel('Year')
plt.ylabel('Growth (%)')
plt.legend()
plt.grid(True)
plt.show()

###So from the above year-over-year graph we can spot that even though there was not enough grwoth in sales in [2014-2015] and [2016-2017] but there was significant growth in profit in those years, whereas when there was a huge growth in sales i.e[2015-2016] in that period there was not much growth in profit